utils/utils.py

```diff
@@ -0,0 +1,299 @@
+import csv
+import os
+import pynini
+from pynini.lib import pynutil,utf8
+import string
+
+def get_abs_path(rel_path):
+    """
+    Get absolute path
+    Args:
+        rel_path: relative path to this file
+        
+    Returns absolute path
+    """
+    return os.path.dirname(os.path.abspath(__file__)) + '/' + rel_path
+
+
+def load_labels(abs_path):
+    """
+    loads relative path file as dictionary
+    Args:
+        abs_path: absolute path
+    Returns dictionary of mappings
+    """
+    label_tsv = open(abs_path, encoding="utf-8")
+    labels = list(csv.reader(label_tsv, delimiter="\t"))
+    return labels
+
+
+DEFAULT_INSERT_COST = 1.
+DEFAULT_DELETE_COST = 1.
+DEFAULT_SUBSTITUTE_COST = 1.
+
+
+class Error(Exception):
+  """Errors specific to this module."""
+
+  pass
+
+
+class EditTransducer:
+  """Factored edit transducer.
+  This class stores the two factors of an finite-alphabet edit transducer and
+  supports insertion, deletion, and substitution operations with user-specified
+  costs.
+  Note that the cost of substitution must be less than the cost of insertion
+  plus the cost of deletion or no optimal path will include substitution.
+  One can impose an upper bound on the number of permissible edits by
+  setting a non-zero value for `bound`. This often results in substantial
+  improvements in performance.
+  """
+
+  # Reserved labels for edit operations.
+  DELETE = "<delete>"
+  INSERT = "<insert>"
+  SUBSTITUTE = "<substitute>"
+
+  def __init__(self,
+               sentence_A: str,
+               sentence_B: str,
+               insert_cost: float = DEFAULT_INSERT_COST,
+               delete_cost: float = DEFAULT_DELETE_COST,
+               substitute_cost: float = DEFAULT_SUBSTITUTE_COST,
+               bound: int = 0):
+    """Constructor.
+    Args:
+      alphabet: edit alphabet (an iterable of strings).
+      insert_cost: the cost for the insertion operation.
+      delete_cost: the cost for the deletion operation.
+      substitute_cost: the cost for the substitution operation.
+      bound: the number of permissible edits, or `0` (the default) if there
+          is no upper bound.
+    """
+    # Left factor; note that we divide the edit costs by two because they also
+    # will be incurred when traversing the right factor.
+    wordlist=[]
+    wordlist_A = sentence_A.split(" ")
+    wordlist_B = sentence_B.split(" ")
+    for word in wordlist_A:
+      if word not in wordlist:
+        strword = '['+ word + ']'
+        wordlist.append(strword)
+    for word in wordlist_B:
+      if word not in wordlist:
+        strword = '[' + word + ']'
+        wordlist.append(strword)
+    sigma = pynini.accep("")
```

Comment:
can we use a single union() to convert the word list? this add-one-by-one is buggy and inefficient